zeroex/orderwatch/order_watcher.go

```diff
@@ -607,10 +607,11 @@ func (w *Watcher) generateOrderEventsIfChanged(hashToOrderWithTxHashes map[commo
 			}
 			orderEvents = append(orderEvents, orderEvent)
 		} else if oldFillableAmount.Cmp(newFillableAmount) == 0 {
-			// No important state-change happened, ignore
-			// Noop
+			// No important state-change happened, simply update lastUpdated timestamp in DB
+			w.updateOrderDBEntry(order)
 		} else if oldFillableAmount.Cmp(big.NewInt(0)) == 1 && oldAmountIsMoreThenNewAmount {
-			// Order was filled, emit  event
+			// Order was filled, emit  event and update order in DB
+			w.updateOrderDBEntry(order)
```

Comment:
@fabioberger I think we need to do this inside of a transaction, otherwise we run the risk of hitting a write-after-read data race. We are trying to update individual fields of the order, but if a separate goroutine is trying to update a field at the same time, one of the changes will be overwritten. So for example:

1. Goroutine-0 reads the order from the database.
2. Goroutine-1 reads the order from the database.
3. Goroutine-0 changes the `FillableTakerAssetAmount` field on its local view of the order.
4. Goroutine-1 changes the `LastUpdated` field on its local view of the order.
5. Goroutine-0 saves the order to the database, including the changes to `FillableTakerAssetAmount`.
6. Goroutine-1 saves the order to the database, overwriting the changes to `FillableTakerAssetAmount`.

This happens because each goroutine has its own local view of the order and is not aware of the changes potentially made by other goroutines.

Unfortunately the only way to avoid this right now is with a lock on the entire `Orders` collection, but hopefully we can take care to hold the lock for as little time as possible. It should be good enough for now. In the future, the `db` package could support (1) locks on individual models instead of entire columns or (2) updating specific fields instead of updating the entire model at once. Either one of those new features would offer a more efficient way to solve the problem. Another way is by using optimistic locking (this is how I [solved the problem in Zoom](https://github.com/albrow/zoom/tree/v0.18.0#concurrent-updates-and-optimistic-locking)).

Reply:
Thanks for finding this issue @albrow. That makes a lot of sense. Looks like we'll need to put the entirety of `handleBlockEvents` within a Orders collection lock. How do I go about acquiring a lock on the Orders collection specifically? I only see the more generic `colLock` and the `globalWriteLock`.

Reply:
Use [`Collection.OpenTransaction`](https://godoc.org/github.com/0xProject/0x-mesh/db#Collection.OpenTransaction).

Reply:
Added the transaction.